pkg/api/util/clusteragent.go

```diff
@@ -0,0 +1,185 @@
+	}
+
+	c.clusterAgentAPIRequestHeaders = &http.Header{}
+	c.clusterAgentAPIRequestHeaders.Set(authorizationHeaderKey, fmt.Sprintf("Bearer %s", authToken))
+
+	// TODO remove insecure
+	c.clusterAgentAPIClient = GetClient(false)
+	c.clusterAgentAPIClient.Timeout = 2 * time.Second
+
+	return nil
+}
+
+// getClusterAgentEndpoint provides a validated https endpoint from:
+// 1. configuration key "cluster_agent_url" like https
+// add the https prefix if the scheme isn't specified
+// 2. environment variables associated with "cluster_agent_kubernetes_service_name"
+// ${dcaServiceName}_SERVICE_HOST and ${dcaServiceName}_SERVICE_PORT
+func getClusterAgentEndpoint() (string, error) {
+	const configDcaURL = "cluster_agent_url"
+	const configDcaSvcName = "cluster_agent_kubernetes_service_name"
+
+	dcaURL := config.Datadog.GetString(configDcaURL)
+	if dcaURL != "" {
+		if strings.HasPrefix(dcaURL, "http://") {
+			return "", fmt.Errorf("cannot get cluster agent endpoint, not a https scheme: %s", dcaURL)
+		}
+		if strings.Contains(dcaURL, "://") == false {
+			log.Tracef("adding https scheme to %s: https://%s", dcaURL, dcaURL)
+			dcaURL = fmt.Sprintf("https://%s", dcaURL)
+		}
+		u, err := url.Parse(dcaURL)
+		if err != nil {
+			return "", err
+		}
+		if u.Scheme != "https" {
+			return "", fmt.Errorf("connot get cluster agent endpoint, not a https scheme: %s", u.Scheme)
+		}
+		return u.String(), nil
+	}
+
+	// Construct the URL with the Kubernetes service environment variables
+	// *_SERVICE_HOST and *_SERVICE_PORT
+	dcaSvc := config.Datadog.GetString(configDcaSvcName)
+	if dcaSvc == "" {
+		return "", fmt.Errorf("cannot get a cluster agent endpoint, both %q and %q are empty", configDcaURL, configDcaSvcName)
+	}
+
+	dcaSvc = strings.ToUpper(dcaSvc)
+
+	// host
+	dcaSvcHostEnv := fmt.Sprintf("%s_SERVICE_HOST", dcaSvc)
+	dcaSvcHost := os.Getenv(dcaSvcHostEnv)
+	if dcaSvcHost == "" {
+		return "", fmt.Errorf("cannot get a cluster agent endpoint for kubernetes service %q, env %q is empty", dcaSvc, dcaSvcHostEnv)
+	}
+
+	// port
+	dcaSvcPort := os.Getenv(fmt.Sprintf("%s_SERVICE_PORT", dcaSvc))
+	if dcaSvcPort == "" {
+		return "", fmt.Errorf("cannot get a cluster agent endpoint for kubernetes service %q, env %q is empty", dcaSvc, dcaSvcPort)
+	}
+
+	// validate the URL
+	dcaURL = fmt.Sprintf("https://%s:%s", dcaSvcHost, dcaSvcPort)
+	u, err := url.Parse(dcaURL)
+	if err != nil {
+		return "", err
+	}
+
+	return u.String(), nil
+}
+
+// GetKubernetesServiceNames query the datadog cluster agent to get nodeName/podName registered
+// Kubernetes services.
+func (c *ClusterAgentUtil) GetKubernetesServiceNames(nodeName, podName string) ([]string, error) {
+	const dcaMetadataPath = "api/v1/metadata"
+	var serviceNames serviceNames
+	var err error
+
+	req := &http.Request{
+		Header: *c.clusterAgentAPIRequestHeaders,
+	}
+	// https://host:port /api/v1/metadata/ {nodeName}/ {pod-[0-9a-z]{5}}
```

Comment:
Not sure about this regex. If I'm not misreading this: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/ it seems like the podname can be much longer